File: Solutions/python/2022/day_03.py
import sys, os
sys.path.append(os.path.abspath(os.path.join(os.path.realpath(os.path.dirname(__file__)), "..", "SetupSolutions")))

from SetupSolutions import getPuzzleInput

year = "2022"
day = "03"
puzzleInput = getPuzzleInput(f"{year}", f"Day_{day}.txt")
puzzle = puzzleInput

def getCommonItems(rucksack_1, rucksack_2):
    common = []
    for letter in rucksack_1:
        if letter in rucksack_2 and letter not in common:
            common.append(letter)
    
    return common

def getCommonItemsThree(rucksack_1, rucksack_2, rucksack_3):
    common = []
    for letter in rucksack_1:
        if letter in rucksack_2 and letter not in common and letter in rucksack_3:
            common.append(letter)
    
    return common

def getItemTypePriority(item):
    value = ord(item)

    if value >= ord("A") and value <= ord("Z"):
        return value - 38
    else:
        return value- 96

def sumOfPrioritiesOfCommonItems(puzzle):
    all_letters = []
    lines = [line for line in puzzle][:3]
    for line in puzzle:
        line = line.replace("\n", "")
        middlePoint = int(len(line) / 2)
        common = getCommonItems(line[0:middlePoint], line[middlePoint:])
        all_letters += [getItemTypePriority(letter) for letter in common]

    return sum(all_letters)

def sumOfPrioritiesOfElfGroups(puzzle):
    all_letters = []
    groups_of_three = [puzzle[i:i + 3] for i in range(0, len(puzzle), 3)]
    for group in groups_of_three:
        common = getCommonItemsThree(group[0].replace("\n", ""), group[1].replace("\n", ""), group[2].replace("\n", ""))
        all_letters += [getItemTypePriority(letter) for letter in common]
    
    return sum(all_letters)
# Script runs here
print(f"Advent of Code - {year}: Day {day}")
print("Part-One Answer:", sumOfPrioritiesOfCommonItems(puzzle))
print("Part-Two Answer:", sumOfPrioritiesOfElfGroups(puzzle))

File: Solutions/javascript/2022/day_03.js
const { getPuzzleInput } = require("../SetupSolutions/SetupSolutions.js");

const year = "2022";
const day = "03";
const puzzleInput = getPuzzleInput(year, `Day_${day}.txt`).filter(el => el);
const puzzle = puzzleInput;

const getCommonItems = (rucksackOne, rucksackTwo) => {
    let common = "";
    for (letter of rucksackOne) {
        if (rucksackTwo.includes(letter) && !common.includes(letter)) {
            common += letter;
        }
    }

    return common;
}

const getCommonItemsThree = (rucksackOne, rucksackTwo, rucksackThree) => {
    let common = "";
    for (letter of rucksackOne) {
        if (rucksackTwo.includes(letter) && rucksackThree.includes(letter) && !common.includes(letter)) {
            common += letter;
        }
    }

    return common;
}

const getItemTypePriority = (item) => {
    const value = item.charCodeAt(0);
    if (value >= "A".charCodeAt(0) && value <= "Z".charCodeAt(0)) {
        return value - 38;
    } else {
        return value - 96;
    }
}

const sumOfPrioritiesOfCommonItems = (puzzle) => {
    let totalCommonItems = [];
    for (line of puzzle) {
        const middlePoint = line.length / 2;
        const sackOne = line.slice(0, middlePoint);
        const sackTwo = line.slice(middlePoint);
        const commonItems = getCommonItems(sackOne, sackTwo);

        totalCommonItems.push(commonItems);
    }

    return totalCommonItems.reduce((prev, curr) => prev + getItemTypePriority(curr), 0);
}

const sumOfPrioritiesOfCommonItemsThreeElves = (puzzle) => {
    let totalCommonItems = [];
    for (let i = 0; i < puzzle.length; i+=3) {
        const commonItems = getCommonItemsThree(puzzle[i], puzzle[i+1], puzzle[i+2]);

        totalCommonItems.push(commonItems);
    }

    return totalCommonItems.reduce((prev, curr) => prev + getItemTypePriority(curr), 0);
}

// Script runs here
console.log(`Advent of Code - ${year}: Day ${day}`);
console.log("Part-One Answer:", sumOfPrioritiesOfCommonItems(puzzle));
console.log("Part-Two Answer:", sumOfPrioritiesOfCommonItemsThreeElves(puzzle));